### ISSUE TITLE
[idea] Changing Parser's language version strategy

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10314]
**The version of ESLint you are using.**

master

**The problem you want to solve.**

Currently, `espree` parses code as ES5 by default. We give `parserOptions.ecmaVersion` option to change that behavior. In this strategy, users have sometimes seen non-understandable error messages such as "unexpected token", and we cannot disable each syntax individually to avoid making the parser more complex.

**Your take on the correct solution to problem.**

1. Make that espree parses code as the latest language version always.
2. ESLint enables the rules which report each syntax like [eslint-plugin-es](https://www.npmjs.com/package/eslint-plugin-es#rules) by default.
3. The environment such as `env: { es2015: true }` disables those syntax rule set.
4. Users can disable each syntax rule individually (e.g. `rules: { "no-async-iteration": "off" }`).

I think that this strategy has some pros:

- We can make more understandable message instead of "unexpected token".
- We can disable each syntax individually.
- We can make the parser simpler by removing version switch.
- We can solve the confusion between `parserOptions.ecmaVersion` and `env.es6`. It's unified to `env`.
- It can detect the use of new built-in's static methods such as `Object.assign`.

On the other hand, this strategy has some cons:

- This is definitely a breaking change that users have to modify their config.
- The breaking changes in new language versions probably break CI builds of users. Though I believe that it's very rare case in order to avoid breaking the Web.
- ~~Adding new rules which report new syntax probably is a breaking change. In that case, it makes hard to add new syntax support.~~ It's OK if the syntax rules are enabled by default only when the `parser` is espree.

**Comment by ilyavolodin** [comment]
I don't think that's a great idea. ESLint is a validation tool, and unintentionally using unsupported language features is something that should be flagged. Rules would help catch those issues, however, a lot of people use shareable configs for rule setup, and if shareable config disables the rule that catches unsupported language features, user might never know about it.

**Comment by not-an-aardvark** [comment]
@ilyavolodin Doesn't the same problem exist now if a shareable config changes the parser options?

**Comment by mysticatea** [comment]
@ilyavolodin I think that there is your concern already since shareable configs can have `parserOptions`.

**Comment by not-an-aardvark** [comment]
I have a few concerns:

* It seems confusing for the rules to only be enabled by default when using `espree` -- we would effectively have a special case for the default parser. If a user decides to fork `espree` or switch to a different parser because they want to add some feature, they might not realize that some rules have been disabled when they switch to their fork.
* It seems like it could cause confusion if these rules are enabled by default, because no other rules are enabled by default.

I think we could improve the "unexpected token" error messages in `espree` separately (or maybe in `acorn`), without needing to make this change.

**Comment by ilyavolodin** [comment]
Good point on `parserOptions`, I forgot that shareable configs can modify those. How would this work with scope analysis? If we always default to latest language version, even if the code is in written in ES5, scopes will be wrong (block scopes for things like `switch` statements, etc.).
I'm still not sold at all on "latest by default" while a lot of new code might be written using something later then ES5, there's a lot of legacy code out there that people have to maintain. I think ES5 by default makes a lot more sense.

**Comment by mysticatea** [comment]
@not-an-aardvark 

> It seems confusing for the rules to only be enabled by default when using espree -- we would effectively have a special case for the default parser. If a user decides to fork espree or switch to a different parser because they want to add some feature, they might not realize that some rules have been disabled when they switch to their fork.

Good point. I thought that the rules which disallow syntaxes don't make sense with custom parsers because the purpose of custom parsers is that introducing new syntax. But I didn't consider about forked espree.

What do you think if we enabled those syntax rules always? My concern is that adding new syntaxes becomes a breaking change since it will enable new rules to disallow new syntaxes.

@platinumazure 

> I would love to see a plugin or shareable config which implements this strategy

The [eslint-plugin-es](https://github.com/mysticatea/eslint-plugin-es#readme) plugin is that one.

Currently, the plugin reports each syntax as just "it's forbidden". But I think that we can add the message "Add `env: {es2018: true}` setting to your .eslintrc file if you want to use this syntax." or something like.

Another sample is [eslint-plugin-node](https://github.com/mysticatea/eslint-plugin-node#readme). The plugin has used the latest `ecmaVersion` and [node/no-unsupported-features](https://github.com/mysticatea/eslint-plugin-node/blob/e957849ce0691794741fb793cab0e61090773340/docs/rules/no-unsupported-features.md) rule while a long time.

@ilyavolodin 

> How would this work with scope analysis? If we always default to latest language version, even if the code is in written in ES5, scopes will be wrong (block scopes for things like switch statements, etc.).
I'm still not sold at all on "latest by default" while a lot of new code might be written using something later then ES5, there's a lot of legacy code out there that people have to maintain. I think ES5 by default makes a lot more sense.

To make sure, my proposal is not that making "latest by default". My proposal is that it changes the way which restrict the target code to ES5 (by default). It's the use of rules which disallow each syntax instead of `parserOptions.ecmaVersion`.

Currently, if users wanted to use ES2018, they write the following setting to opt in ES2018:

```js
{
    "parserOptions": {
        "ecmaVersion": 2018 // enable ES2018 syntax
    },
    "env": {
        "es6": true // enable ES2015 globals
    },
    "globals": {
        "Atomics": false, // enable ES2017 globals
        "SharedArrayBuffer": false // enable ES2017 globals
    }
}
```

(it's a confusion for newcomers: why are many separated settings needed?)

After this proposal, they write the following setting to opt out rules which disallow ES2018:

```js
{
    "env": {
        "es2018": true // disable the rules which disallow ES2015-ES2018
    }
}
```

Don't worry about ES5. ES2015 don't change existing syntax's behaviors to avoid breaking the Web. For example, it added `let`/`const` declarations to use block-scoped varaibles, but it didn't change `var` declarations to block-scoped. The `var` declarations are still function-scoped. The `eslint-scope` handles it correctly.

Plus, this new strategy can handle ES5 more properly. For example, the current ESLint doesn't catch the use of `Object.assign` that ES2015 added. On the other hand, this new strategy can catch it by the rule such as [es/no-object-assign](https://github.com/mysticatea/eslint-plugin-es/blob/d52f9cfcabfb47cb4093311e546b6adcba97dd81/docs/rules/no-object-assign.md).

----

I don't think that we are hurry to decide about this since this is a breaking change and 5.0.0 doesn't include.
